fairlearn/reductions/_moments/cdf_demographic_parity_moment.py

```diff
@@ -0,0 +1,198 @@
+        `AbsoluteLoss`.
+    y_range : (float,float)
+        A tiple that specifies the range of y labels of the data.
+    differece_bound : bool
+        The constraints' difference bound for constraints that are expressed
+        as differences, also referred to as :math:`\\epsilon` in documentation.
+        Default None
+    grids: list of floats
+        Instead of supplying a number for the grid, users may specify the exact
+        set of grids they desire using this argument.
+    grid_num: int
+        If use does not specify grids, then grid_num and y_range is used to calculate the grids
+        used as thresholds.
+    """
+    short_name = "CDF_DemographicParity"
+
+    def __init__(self, loss, y_range, difference_bound=None, grids=None, grid_num=41):
+        super().__init__()
+        self.utility_parity = UtilityParity(difference_bound=difference_bound)
+        self.loss = loss
+        self.objective = WeightedErrorRate()
+        self.grids = grids
+        self.grid_num = grid_num
+        self.y_range = y_range
+
+        # check the validity of y_range
+        assert (self.y_range is not None), "You should specify the range of y"
+        assert self.y_range[0] <= self.y_range[1], "The range of the label y is not valid!"
+
+        # determine the grids
+        if self.grids == None:
+            self.grids = np.linspace(self.y_range[0], self.y_range[1], self.grid_num)
+        else:
+            # check whether all grids fall into the empirical range of y
+            assert (self.grids <= self.y_range[1]).all()\
+                    and (self.grids >= self.y_range[0]).all(),\
+                    "Some of the grids do not fall into the range of y"
+            self.grids = np.unique(self.grids)
+            # ensure grids are listed in an increasing order and are unique
+
+        # in case the grids have already included the two endpoints of the range
+        # prepare for the calculation of optimal labels
+        self.pred_grid = [self.y_range[0]] +\
+            list([self.next_grid(grid) for grid in self.grids]) + [self.y_range[1]]
+        self.pred_grid = list(filter(lambda x: x >= self.y_range[0], self.pred_grid))
+        self.pred_grid = list(filter(lambda x: x <= self.y_range[1], self.pred_grid))
+        self.pred_vec = {}
+        for pred in self.pred_grid:
+            self.pred_vec[pred] = (1 * (pred >= pd.Series(self.grids)))
+
+    def prev_grid(self, theta):
+        index = np.where(self.grids == theta)[0][0]
+        if index == 0:
+            return self.grids[index] - (self.grids[1]-self.grids[0])/2
+        else:
+            return (self.grids[index] + self.grids[index-1])/2
+
+    def next_grid(self, theta):
+        index = np.where(self.grids == theta)[0][0]
+        if index + 1 == self.grids.size:
+            return self.grids[index] + (self.grids[1]-self.grids[0])/2
+        else:
+            return (self.grids[index+1] + self.grids[index])/2
+
+    def augment_data(self, x, a, y):
```

Comment:
Did you mean capital X?
Is `a` meant to be `sensitive_features`? If so, please replace.